### ISSUE TITLE
[AMD][kernel][Tier 6] Port sol-execbench L2/056_language_model_decoder_prenorm_attention_ffn_residual_backward to FlyDSL

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/FlyDSL#229]
# FlyDSL implementation: port sol-execbench's `056_language_model_decoder_prenorm_attention_ffn_residual_backward` problem to FlyDSL

## Goal

Port the `056_language_model_decoder_prenorm_attention_ffn_residual_backward` problem from `nvidia/SOL-ExecBench` (subset `L2`) to FlyDSL as part of the sol-execbench → FlyDSL coverage sweep. This is one of 235 problems across 4 subsets (L1, L2, Quant, FlashInfer-Bench) being tracked for FlyDSL kernel coverage.

**Source:** `mistralai/Voxtral-Small-24B-2507` (per `definition.json` `hf_id`)

**Description:** Backward pass for complete decoder transformer layer with pre-norm attention and FFN with residual connections. Computes gradients for all weights and input through RMSNorm, GQA attention with RoPE, and SwiGLU FFN.

## Hardware / workload

```text
Hardware:  1 × MI300X (gfx942) + 1 × MI355X (gfx950)
Models:    (none — kernel benchmark, not LLM workload)
Subset:    L2
Problem:   056_language_model_decoder_prenorm_attention_ffn_residual_backward
Source:    mistralai/Voxtral-Small-24B-2507
```

**Const axes:**
  - `hidden_size` = 5120
  - `num_heads` = 32
  - `num_kv_heads` = 8
  - `head_dim` = 160
  - `intermediate_size` = 14336

**Inputs:**
  - `grad_output` [batch_size, seq_len, hidden_size] bfloat16 — Gradient from upstream layer
  - `residual` [batch_size, seq_len, hidden_size] bfloat16 — Original input (residual connection)
  - `attn_input` [batch_size, seq_len, hidden_size] bfloat16 — Input to attention after first layernorm
  - `query_states` [batch_size, num_heads, seq_len, head_dim] bfloat16 — Query states before RoPE
  - `key_states` [batch_size, num_kv_heads, seq_len, head_dim] bfloat16 — Key states before RoPE
  - `value_states` [batch_size, num_kv_heads, seq_len, head_dim] bfloat16 — Value states
  - `query_states_rotated` [batch_size, num_heads, seq_len, head_dim] bfloat16 — Query states after RoPE
  - `key_states_rotated` [batch_size, num_kv_heads, seq_len, head_dim] bfloat16 — Key states after RoPE
  - `key_states_repeated` [batch_size, num_heads, seq_len, head_dim] bfloat16 — Key states repeated for GQA
  - `value_states_repeated` [batch_size, num_heads, seq_len, head_dim] bfloat16 — Value states repeated for GQA
  - `cos` [batch_size, seq_len, head_dim] bfloat16 — Cosine for RoPE
  - `sin` [batch_size, seq_len, head_dim] bfloat16 — Sine for RoPE
  - `attn_weights` [batch_size, num_heads, seq_len, seq_len] bfloat16 — Attention weights after softmax
  - `attn_output` [batch_size, seq_len, num_heads_x_head_dim] bfloat16 — Attention output before projection
  - `residual2` [batch_size, seq_len, hidden_size] bfloat16 — Input to FFN block (residual connection)
  - `ffn_input` [batch_size, seq_len, hidden_size] bfloat16 — Input to FFN after second layernorm
  - `gate` [batch_size, seq_len, intermediate_size] bfloat16 — Gate projection output
  - `up` [batch_size, seq_len, intermediate_size] bfloat16 — Up projection output
  - `silu_up` [batch_size, seq_len, intermediate_size] bfloat16 — SiLU activation of up projection
  - `swiglu_output` [batch_size, seq_len, intermediate_size] bfloat16 — SwiGLU output (gate * silu_up)
  - `input_ln_weight` [hidden_size] bfloat16 — Input layernorm weight
  - `q_weight` [num_heads_x_head_dim, hidden_size] bfloat16 — Query projection weight
  - `k_weight` [num_kv_heads_x_head_dim, hidden_size] bfloat16 — Key projection weight
  - `v_weight` [num_kv_heads_x_head_dim, hidden_size] bfloat16 — Value projection weight
  - `o_weight` [hidden_size, num_heads_x_head_dim] bfloat16 — Output projection weight
  - `post_attn_ln_weight` [hidden_size] bfloat16 — Post-attention layernorm weight
  - `gate_weight` [intermediate_size, hidden_size] bfloat16 — Gate projection weight
  - `up_weight` [intermediate_size, hidden_size] bfloat16 — Up projection weight
  - `down_weight` [hidden_size, intermediate_size] bfloat16 — Down projection weight
  - `variance1` [batch_size, seq_len, 1] float32 — Variance from first RMSNorm
  - `variance2` [batch_size, seq_len, 1] float32 — Variance from second RMSNorm
  - `hidden_states_normalized1` [batch_size, seq_len, hidden_size] float32 — Normalized hidden states from first RMSNorm
  - `hidden_states_normalized2` [batch_size, seq_len, hidden_size] float32 — Normalized hidden states from second RMSNorm
  - `eps` [] float32 — Epsilon for numerical stability

**Outputs:**
  - `grad_input` [batch_size, seq_len, hidden_size] bfloat16 — Gradient w.r.t. input hidden states
  - `grad_input_ln_weight` [hidden_size] float32 — Gradient w.r.t. input layernorm weight
  - `grad_q_weight` [num_heads_x_head_dim, hidden_size] bfloat16 — Gradient w.r.t. query projection weight
  - `grad_k_weight` [num_kv_heads_x_head_dim, hidden_size] bfloat16 — Gradient w.r.t. key projection weight
  - `grad_v_weight` [num_kv_heads_x_head_dim, hidden_size] bfloat16 — Gradient w.r.t. value projection weight
  - `grad_o_weight` [hidden_size, num_heads_x_head_dim] bfloat16 — Gradient w.r.t. output projection weight
  - `grad_post_attn_ln_weight` [hidden_size] float32 — Gradient w.r.t. post-attention layernorm weight
  - `grad_gate_weight` [intermediate_size, hidden_size] bfloat16 — Gradient w.r.t. gate projection weight
  - `grad_up_weight` [intermediate_size, hidden_size] bfloat16 — Gradient w.r.t. up projection weight
  - `grad_down_weight` [hidden_size, intermediate_size] bfloat16 — Gradient w.r.t. down projection weight

**Workload samples (first 3):**
  - {"uuid": "2a80ae77-d5f7-5f70-aca4-2b61b067a5c2", "axes": {"batch_size": 1, "seq_len": 256}, "inputs": {"grad_output": {"type": "custom"}, "residual": {"type": "custom"}, "attn_input": {"type": "custom
  - {"uuid": "7f5e72f9-91d9-554e-ac35-4ce5c1ff5147", "axes": {"batch_size": 4, "seq_len": 293}, "inputs": {"grad_output": {"type": "custom"}, "residual": {"type": "custom"}, "attn_input": {"type": "custom
  - {"uuid": "e476a7b8-7a16-55e8-aa5c-2c0a49b81686", "axes": {"batch_size": 8, "seq_len": 256}, "inputs": {"grad_output": {"type": "custom"}, "residual": {"type": "custom"}, "attn_input": {"type": "custom

```text
Baseline:  sol-execbench PyTorch reference (reference.py) on AMD with torch.compile off
Target:    max_abs_diff < 1e-2 vs PyTorch reference on all tested shapes
           kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
```

## Scope

1. `kernels/sol_execbench/l2_056_language_model_decoder_prenorm_attention_ffn_residual_backward.py` — FlyDSL kernel implementing the op described in `definition.json`, matching the PyTorch reference's numerics
2. `tests/kernels/test_sol_execbench_l2_056_language_model_decoder_prenorm_attention_ffn_residual_backward.py` — parity harness vs PyTorch reference, testing multiple workload shapes
3. `tests/kernels/bench_sol_execbench_l2_056_language_model_decoder_prenorm_attention_ffn_residual_backward.py` — reports `kernel_ms` per shape + comparison vs PyTorch eager baseline
4. `assessment.md` (≤300 words) — ranking which shapes FlyDSL wins/ties/loses vs PyTorch eager

## Acceptance

- Correctness: max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes
- Performance: kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
- Coverage: works on both gfx942 (MI300X) and gfx950 (MI355X)

## Out of scope

- Backward pass (unless the problem is itself a backward op)
- Weight gradients
- Optimizer step
- Cross-arch comparison vs NVIDIA H100 (PyTorch eager on AMD is the baseline)

## Reproduction

Clone `amdpilot-org/SOL-ExecBench`, download the dataset (`scripts/download_data.sh`), find the problem at `data/benchmark/L2/056_language_model_decoder_prenorm_attention_ffn_residual_backward/`, run the sol-execbench driver against a FlyDSL solution. Requires the unified aiter+triton+flydsl Docker image at `/home/xiasun/unified-rocm.Dockerfile` (not yet built).

## Definition of done

- [ ] Kernel implemented at `kernels/sol_execbench/l2_056_language_model_decoder_prenorm_attention_ffn_residual_backward.py`
- [ ] Tests pass on MI300X (gfx942)
- [ ] Tests pass on MI355X (gfx950)
- [ ] Bench shows ≥80% win rate vs PyTorch eager
- [ ] `assessment.md` written

## ## Leaderboard baseline to beat

These are the current best public results on NVIDIA B200 (from the
[sol-execbench leaderboard](https://research.nvidia.com/benchmarks/sol-execbench/leaderboard))
for problem `056_language_model_decoder_prenorm_attention_ffn_residual_backward` (kernel_id=150). The agent's
FlyDSL kernel on MI300X/MI355X should aim to beat the **Scoring Baseline**
(optimized PyTorch) and approach the **top human/AI submissions**, normalized
by cross-arch roofline.

```text
GPU:           NVIDIA B200 (leaderboard) vs AMD MI300X/MI355X (this task)
Latency unit:  ms

SOL Bound (theoretical roofline):     0.810574 ms   (SOL-Score=1.0000, speedup=2.62×)
Scoring Baseline (optimized PyTorch): 2.102038 ms   (SOL-Score=0.4872, speedup=0.97×)
Reference Implementation (plain PyTorch): 3.360605 ms   (SOL-Score=0.3158, speedup=0.60×)

Top public submissions:
  - **Rank 1** (`Proud Wren @ NVIDIA`): 1.499085 ms, SOL-Score=0.6424, speedup=1.35× vs reference
  - **Rank 2** (`Hyra`): 1.817130 ms, SOL-Score=0.5476
  - **Rank 3** (`Amir M. Mir | SF Tensor`): 1.892880 ms, SOL-Score=0.5293
```

**Cross-arch note:** B200 timings are not directly comparable to MI300X/MI355X
(different GPU arch, different roofline). The agent should:
1. Measure the PyTorch reference on MI300X/MI355X (same shapes, same Docker image)
2. Compute the FlyDSL kernel's speedup vs the on-AMD PyTorch reference
3. Compare that speedup ratio to the leaderboard's `avg_speedup` ratios above
4. Target: FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × (Rank-1 speedup vs PyTorch-reference-on-B200)

## Definition of done

- [ ] FlyDSL kernel passes correctness (max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes)
- [ ] FlyDSL kernel_ms measured on MI300X (gfx942) AND MI355X (gfx950) across >=5 workload shapes
- [ ] FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × Rank-1 speedup vs PyTorch-reference-on-B200 (=1.35× if top1 exists)
- [ ] `assessment.md` (<=300 words) ranks FlyDSL vs PyTorch eager on each shape, identifies win/tie/lose regimes
- [ ] Patch modifies actual kernel source (not just bench scripts) — bench-only patches are REJECTED

## Non-goals / rejection criteria

- Bench-only patches (no FlyDSL kernel source changes) are rejected
- Patches that set expected_value = measured_value (tautological pass) are rejected
- The metric contract must compare FlyDSL vs PyTorch reference on AMD, not an absolute ms
- "Completed" without a measured speedup ratio vs the on-AMD PyTorch reference is rejected

**Comment by sunxxuns** [comment]
## AMDPilot acceptance target

**The deliverable is a FlyDSL kernel.** A PyTorch reimplementation of this
problem -- with or without `torch.compile` -- is the *baseline*, not a result.
It does not close this issue no matter how closely it matches the reference.

### Implementation contract (hard requirement)

The submitted kernel MUST be written in FlyDSL and go through the FlyDSL
compiler. Concretely, the kernel module has to actually use the DSL, e.g.:

```python
import flydsl.compiler as flyc
from flydsl.expr import arith as ea, gpu, vector as ev
```

and dispatch through `flyc.jit` / `flyc.kernel`. A file placed under
`FlyDSL/kernels/sol_execbench/` that only calls `torch.*` ops does NOT satisfy
this issue, even if it is numerically perfect and even if it is faster.

If you conclude the port is not expressible in FlyDSL today, say so explicitly
and report exactly which DSL capability is missing. That is a legitimate,
useful outcome. Silently substituting a PyTorch implementation is not.

### Baseline definition

The baseline is the sol-execbench PyTorch reference (`reference.py`) running
eager, with `torch.compile` OFF. Measure it yourself in this run, on the same
GPU and the same workload shapes you benchmark your kernel on. Do not carry a
number over from another run, another GPU, or the issue text.

### Performance target

Your FlyDSL kernel must be **at least 10% faster than that measured baseline**
-- `kernel_ms <= 0.90 x baseline_ms` -- on **at least 80% of the tested
workload shapes**.

### Correctness target

`max_abs_diff < 1e-2` versus the PyTorch reference on **all** tested shapes. A
faster kernel that fails this is a failure, not a tradeoff.

### What you must report

Report a per-shape table containing, for every tested shape: the measured
eager baseline ms, your FlyDSL kernel ms, the ratio, and win/loss. Then state
the win rate against the 80% bar. A result without the baseline column next to
the kernel column is not checkable and will not be credited.
